---
title: Pharmacogenomics Prediction Pipeline (P3)
output:
  knitrBootstrap::bootstrap_document:
    theme: flatly
---

<div style='font-size:36px';>Pharmacogenomics Prediction Pipeline (P3) Data Visualization</div>

Overview
========

The purpose of this file is to explore some of the basic properties of the
various raw and processed data files used in this analysis. This may be helpful
to catch artifacts and outliers in the data as well as to detect possible
trends and groupings within the data.

Setup
=====

```{r knitr_settings, include=FALSE}
library('knitr')
opts_knit$set(progress=FALSE, verbose=TRUE)
opts_chunk$set(fig.width=1920/96,
               fig.height=1920/96,
               dpi=96)
options(digits=4)
options(stringsAsFactors=FALSE)
options(knitr.duplicate.label='allow')

rm(list=ls())    # Clean up any existing variables
```


```{r load_libraries}
library('corpcor')
library('dplyr')
library('gplots')
library('ggplot2')
library('hpgltools')
library('RColorBrewer')
```

```{r helper_functions}
#' Compute singular value decomposition
#'
#' https://github.com/kokrah/cbcbSEQ/
#'
#' @param x matrix of genes by sample (ie. the usual data matrix)
#' @return returns a list of svd components v and d
#' @export
makeSVD = function(x){
  x = as.matrix(x)
  s = fast.svd(x - rowMeans(x))
  
  v = s$v
  rownames(v) = colnames(x)
  
  s = list(v=v, d=s$d)
  return(s)
}

#
# Counts-per-million (CPM)
#
cpm = function (x) {
    sweep(x, 2, colSums(x), '/') * 1E6
}

#
# Data we need from the AWS instance
#
get_data = function(filepath){
    # e.g., "ubuntu@123.45.678"
    host = Sys.getenv("HOSTSTRING")
    dir.create(dirname(filepath), recursive=TRUE, showWarnings=FALSE)
    system(paste0("rsync -avr --progress ", host, ':', filepath, ' ', filepath))
}
```

Raw Data
========

RNA-Seq
-------

```{r download_raw_counts}
rnaseq_input = '/data/datasets/raw/rnaseq_expression/HMCL_ensembl74_Counts.csv'
lapply(rnaseq_input, get_data)
```

```{r load_rnaseq}
raw_counts = as.matrix(read.csv(rnaseq_input, header=1, row.names=1))

# Cell line metadata
# http://www.keatslab.org/projects/mm-cell-line-characterization/cell-line-characterization-status
il6_dependent = c('OH2_PLB', 'ALMC2_DJ', 'PE1_PLB', 'ANBL6_DJ2', 'PCM6_Riken', 'ALMC1_DJ')

sample_metadata = data.frame(
    cell_line=colnames(raw_counts),
    il6_dependent=colnames(raw_counts) %in% il6_dependent
)

# color by IL-6 inclusion
il6_colors = ifelse(sample_metadata$il6_dependent, 'red', 'blue')
```

### Sample sizes

```{r rnaseq_sample_sizes}
hpgl_libsize(raw_counts, design=design, scale=FALSE)
```

### Cell line trends

#### PCA

```{r cell_lines}
svd_result = makeSVD(raw_counts)

pc_variance = round((svd_result$d^2) / sum(svd_result$d^2) * 100, 2)
xl = sprintf("PC1 (%.2f%% variance)", pc_variance[1])
yl = sprintf("PC2 (%.2f%% variance)", pc_variance[2])

pca_data = data.frame(PC1=svd_result$v[,1], PC2=svd_result$v[,2],
                      il6=il6_colors,
                      cell_line=sample_metadata$cell_line)

plt = ggplot(pca_data, aes(PC1, PC2, color=il6)) +
    geom_point(stat="identity") +
    geom_text(aes(label=cell_line), angle=45, size=4,vjust=2) +
    xlab(xl) + ylab(yl) +
    ggtitle(sprintf("PCA: RNA-Seq")) +
    theme(axis.ticks=element_blank(), axis.text.x=element_text(angle=-90))
plot(plt)
```

#### Heatmap (Raw counts)

```{r raw_counts_heatmap}
plot(density(raw_counts))

heatmap.2(raw_counts, trace="none", dendrogram='column', labRow=NA,
          col=redgreen(75), colRow=NA, main="Raw RNA-Seq counts",
          ColSideColors=il6_colors, margins=c(12,8))
```

#### Heatmap (Cell line correlation)

```{r rnaseq_heatmaps}
# RNA-Seq Cell line heatmap
heatmap.2(cor(raw_counts), trace="none",
          col=redgreen(75), colRow=NA, main="Cell line RNA-Seq (Raw)",
          ColSideColors=il6_colors,
          margins=c(12,8))

# RNA-Seq Cell line heatmap (log2-CPM)
log2cpm_counts = log2(cpm(raw_counts) + 0.5)
heatmap.2(cor(log2cpm_counts), trace="none",
          col=redgreen(75), labRow=NA, main="Cell line RNA-Seq (log2-CPM)",
          ColSideColors=il6_colors,
          margins=c(12,8))
```

Drug-response
-------------

Drug-response profiles.

```{r download_raw_drug_response_data}
drug_response_input = '/data/datasets/raw/drug_response/NCATS_DxData.txt'
lapply(drug_response_input, get_data)
```

```{r load_drug_response_data}
dx = read.delim(drug_response_input)
dx$drug = rownames(dx)
dx = tbl_df(dx)
```

### PCA

```{r cell_line}
dx_data = as.matrix(dx %>% select(starts_with('DATA'), TAUC))

# PCA
svd_result = makeSVD(t(dx_data))

pc_variance = round((svd_result$d^2) / sum(svd_result$d^2) * 100, 2)
xl = sprintf("PC1 (%.2f%% variance)", pc_variance[1])
yl = sprintf("PC2 (%.2f%% variance)", pc_variance[2])

pca_data = data.frame(drug=rownames(dx),
                      PC1=svd_result$v[,1], PC2=svd_result$v[,2],
                      target=dx$TARGET)

plt = ggplot(pca_data, aes(PC1, PC2, color=target)) +
    geom_point(stat="identity") +
    xlab(xl) + ylab(yl) +
    ggtitle(sprintf("PCA: Drug Response")) +
    theme(axis.ticks=element_blank(), axis.text.x=element_text(angle=-90))
plot(plt)

# Heatmap
#heatmap.2(cor(dx_data), trace="none",
#          col=redgreen(75), labRow=NA, main="Drug response profiles",
#          margins=c(12,8))
```

Filtered Data
=============

Normalized RNA-Seq
------------------

RNA-Seq samples have been processed to:

1. Remove outlier samples
2. Remove low-count genes
3. CPM transform
4. Log2 transform
5. Quantile normalize

```{r download_normed_counts}
rnaseq_input = '/data/datasets/filtered/rnaseq_expression/HMCL_ensembl74_Counts_normalized.csv'
rnaseq_zscores = '/data/datasets/filtered/rnaseq_expression/HMCL_ensembl74_Counts_zscore.csv'
lapply(c(rnaseq_input, rnaseq_zscores), get_data)
```

### Heatmap

```{r normalized_rnaseq}
normed_counts = as.matrix(read.csv(rnaseq_input, row.names=1))

# Sample metadata 
sample_metadata = data.frame(
    cell_line=colnames(normed_counts),
    il6_dependent=colnames(normed_counts) %in% il6_dependent
)

# color by IL-6 inclusion
il6_colors = ifelse(sample_metadata$il6_dependent, 'red', 'blue')

# RNA-Seq Cell line heatmap (Normalized with low-count genes filtered)
heatmap.2(cor(normed_counts), trace="none",
          col=redgreen(75), colRow=NA, main="Cell line RNA-Seq (Filtered & Normalized)",
          ColSideColors=il6_colors,
          margins=c(12,8))
```

Cell line differential expression (Z-scores)
--------------------------------------------

#### Heatmap

```{r rnaseq_zscores, fig.width=1080/96, fig.height=1920/96}
# RNA-Seq z-scores
zscores = as.matrix(read.csv(rnaseq_zscores, row.names=1))

plot(density(zscores))

# Clip outliers to improve color scale
zscores_clipped = pmin(pmax(zscores, -10), 10)

# Clipped z-scores
plot(density(zscores_clipped))

heatmap.2(zscores_clipped, trace="none",
          dendrogram='column', labRow=NA,
          col=redgreen(75), colRow=NA, main="Normalized RNA-Seq Z-scores",
          ColSideColors=il6_colors,
          margins=c(12,8))
```

#### PCA

```{r rnaseq_zscores_pca}
svd_result = makeSVD(zscores)

pc_variance = round((svd_result$d^2) / sum(svd_result$d^2) * 100, 2)
xl = sprintf("PC1 (%.2f%% variance)", pc_variance[1])
yl = sprintf("PC2 (%.2f%% variance)", pc_variance[2])

pca_data = data.frame(PC1=svd_result$v[,1], PC2=svd_result$v[,2],
                      il6=il6_colors,
                      cell_line=sample_metadata$cell_line)

plt = ggplot(pca_data, aes(PC1, PC2, color=il6)) +
    geom_point(stat="identity") +
    geom_text(aes(label=cell_line), angle=45, size=4,vjust=2) +
    xlab(xl) + ylab(yl) +
    ggtitle(sprintf("PCA: RNA-Seq Z-scores")) +
    theme(axis.ticks=element_blank(), axis.text.x=element_text(angle=-90))
plot(plt)
```

Drug response AC-50s
--------------------

```{r drug_response_ac50}
infile  = "/data/datasets/filtered/drug_response/iAC50_filtered.csv"
boxplot(ngs_table) 
boxplot(t(ngs_table))
```

System information
==================

```{r sysinfo}
sessionInfo()
```

